### PR TITLE
When rendering into secondary views, copy the pixel contents between contexts

### PR DESCRIPTION
**Pull request by robert-ancell**
This is because GTK3 can't share contexts between GtkGLAreas.

We may be able to avoid the copy using the GL_OES_EGL_image or GL_OES_EGL_image_external extensions.

Note that keyboard input doesn't work on the secondary view, this requires moving the keyboard handling from FlView to FlEngine. Proposing this now as a step forwards, but not yet useful for real applications.

### CODE REVIEW

**Comment by robert-ancell** [comment]
You can test this with the following Flutter application:

`main.dart`:
```dart
import 'dart:ui' show FlutterView;
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

MethodChannel viewChannel = const MethodChannel('flutter/view');

void main() {
  runWidget(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => MyAppState();
}

class MyAppState extends State<MyApp> with WidgetsBindingObserver {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangeMetrics() {
    print('didChangeMetrics');
    _refresh();
  }

  @override
  Widget build(BuildContext context) {
    var views = <Widget>[];
    for (final FlutterView view
        in WidgetsBinding.instance.platformDispatcher.views) {
      views.add(
        View(
          view: view,
          child: Directionality(
            textDirection: TextDirection.ltr,
            child: Container(
              color: Colors.grey,
              child: Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Text(
                        'View#${view.viewId} ${view.physicalConstraints.maxWidth}x${view.physicalConstraints.maxHeight}@${view.devicePixelRatio}'),
		    SizedBox(height: 10),
		    SizedBox(width: 100, child: EditableText(controller: TextEditingController(), focusNode: FocusNode(), style: TextStyle(), cursorColor: Colors.black, backgroundCursorColor: Colors.black),),
                    if (view.viewId == 0) SizedBox(height: 10),
                    if (view.viewId == 0)
                      OutlinedButton(
                        onPressed: _addView,
                        child: Text('Add Window'),
                      ),
                  ],
                ),
              ),
            ),
          ),
        ),
      );
    }
    return ViewCollection(views: views);
  }

  void _refresh() {
    setState(() {});
  }

  void _addView() {
    viewChannel.invokeMethod('Add');
  }
}
```

`my_application.cc`:
```c
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;

  FlView *implicit_view;

  FlMethodChannel *view_channel;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

static FlMethodResponse *add_view(MyApplication *self,
                                  FlMethodCall *method_call) {
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(self)));
  gtk_window_set_default_size(window, 320, 240);
  gtk_widget_show(GTK_WIDGET(window));

  FlEngine *engine = fl_view_get_engine(self->implicit_view);
  FlView *view = fl_view_new_for_engine(engine);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  return FL_METHOD_RESPONSE(fl_method_success_response_new(nullptr));
}

// Sends the method call response to Flutter.
static void send_response(FlMethodCall *method_call,
                          FlMethodResponse *response) {
  g_autoptr(GError) error = nullptr;
  if (!fl_method_call_respond(method_call, response, &error)) {
    g_warning("Failed to send method call response: %s", error->message);
  }
}

static void view_method_call_cb(FlMethodChannel *channel,
                                FlMethodCall *method_call, gpointer user_data) {
  MyApplication *self = MY_APPLICATION(user_data);

  const gchar *method = fl_method_call_get_name(method_call);

  g_autoptr(FlMethodResponse) response = nullptr;
  if (strcmp(method, "Add") == 0) {
    response = add_view(self, method_call);
  } else {
    response = FL_METHOD_RESPONSE(fl_method_not_implemented_response_new());
  }

  if (response != nullptr) {
    send_response(method_call, response);
  }
}

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));
  gtk_window_set_default_size(window, 640, 480);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  self->implicit_view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(self->implicit_view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(self->implicit_view));

  g_autoptr(FlStandardMethodCodec) codec = fl_standard_method_codec_new();
  FlEngine *engine = fl_view_get_engine(self->implicit_view);
  self->view_channel =
      fl_method_channel_new(fl_engine_get_binary_messenger(engine),
                            "flutter/view", FL_METHOD_CODEC(codec));
  fl_method_channel_set_method_call_handler(self->view_channel,
                                            view_method_call_cb, self, nullptr);

  fl_register_plugins(FL_PLUGIN_REGISTRY(self->implicit_view));

  gtk_widget_grab_focus(GTK_WIDGET(self->implicit_view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GApplication::startup.
static void my_application_startup(GApplication* application) {
  //MyApplication* self = MY_APPLICATION(object);

  // Perform any actions required at application startup.

  G_APPLICATION_CLASS(my_application_parent_class)->startup(application);
}

// Implements GApplication::shutdown.
static void my_application_shutdown(GApplication* application) {
  //MyApplication* self = MY_APPLICATION(object);

  // Perform any actions required at application shutdown.

  G_APPLICATION_CLASS(my_application_parent_class)->shutdown(application);
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  g_clear_object(&self->view_channel);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_APPLICATION_CLASS(klass)->startup = my_application_startup;
  G_APPLICATION_CLASS(klass)->shutdown = my_application_shutdown;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  // Set the program name to the application ID, which helps various systems
  // like GTK and desktop environments map this running application to its
  // corresponding .desktop file. This ensures better integration by allowing
  // the application to be recognized beyond its binary name.
  g_set_prgname(APPLICATION_ID);

  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}
```

File: shell/platform/linux/fl_renderer.cc [hunk 1]
```diff
@@ -474,10 +483,73 @@ gboolean fl_renderer_present_layers(FlRenderer* self,
...
+          width = w;
+        }
+        if (h > height) {
+          height = h;
+        }
+      }
+
+      FlFramebuffer* view_framebuffer =
+          fl_framebuffer_new(priv->general_format, width, height);
+      glBindFramebuffer(GL_DRAW_FRAMEBUFFER,
```
Comment (loic-sharma) [inline shell/platform/linux/fl_renderer.cc:515]: What's Flutter Linux's best practices around OpenGL error reporting? Should we be checking for errors after GL calls?

On Flutter Windows we started migrating to Impeller's OpenGL abstractions. Specifically, we use their [GLES proc table](https://github.com/flutter/engine/blob/main/impeller/renderer/backend/gles/proc_table_gles.h) to get conveniences like automatic error reporting. Here's a Flutter Windows [example](https://github.com/flutter/engine/blob/9513c54456927336666df817a0fa13882ea5faea/shell/platform/windows/compositor_opengl.cc#L39-L52). This has been really useful for debugging weird GL errors. In the future, should Flutter Linux move to this as well?

**Review by loic-sharma — APPROVED** [review]
This looks good to me - thanks!

File: shell/platform/linux/fl_renderer.cc (hunk 1 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_renderer.cc:515]: You can turn on debugging using the `GDK_DEBUG=opengl` environment variable (this is built into GTK). I think this is sufficient if you are aware of any issue and are investigating.

**Comment by robert-ancell** [comment]
Created https://github.com/flutter/engine/pull/55763 to make this easier to test.

**Comment by robert-ancell** [comment]
This branch now has tests, but I'll land https://github.com/flutter/engine/pull/55763 first to make the git history nicer.

**Pull request merged**